libpod/runtime.go

```diff
@@ -1,26 +1,135 @@
+	config          *RuntimeConfig
+	store           storage.Store
+	imageContext    *types.SystemContext
+	apparmorEnabled bool
+	seccompEnabled  bool
+	valid           bool
+	lock            sync.RWMutex
+}
+
+// RuntimeConfig contains configuration options used to set up the runtime
+type RuntimeConfig struct {
+	StorageConfig         storage.StoreOptions
+	ImageDefaultTransport string
+	InsecureRegistries    []string
+	Registries            []string
+	SignaturePolicyPath   string
+	RuntimePath           string
+	ConmonPath            string
+	ConmonEnvVars         []string
+	CgroupManager         string
+	SelinuxEnabled        bool
+	PidsLimit             int64
 }
 
+var (
+	defaultRuntimeConfig = RuntimeConfig{
+		// Leave this empty so containers/storage will use its defaults
+		StorageConfig:         storage.StoreOptions{},
+		ImageDefaultTransport: "docker://",
+		RuntimePath:           "/usr/bin/runc",
+		ConmonPath:            "/usr/local/libexec/crio/conmon",
+		ConmonEnvVars: []string{
+			"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+		},
+		CgroupManager:  "cgroupfs",
+		SelinuxEnabled: false,
+		PidsLimit:      1024,
+	}
+)
+
 // NewRuntime creates a new container runtime
+// Options can be passed to override the default configuration for the runtime
 func NewRuntime(options ...RuntimeOption) (*Runtime, error) {
-	return nil, ctr.ErrNotImplemented
+	runtime := new(Runtime)
+	runtime.config = new(RuntimeConfig)
+
+	// Copy the default configuration
+	deepcopier.Copy(defaultRuntimeConfig).To(runtime.config)
+
+	// Overwrite it with user-given configuration options
+	for _, opt := range options {
+		if err := opt(runtime); err != nil {
+			return nil, errors.Wrapf(err, "error configuring runtime")
+		}
+	}
+
+	// Set up containers/storage
+	store, err := storage.GetStore(runtime.config.StorageConfig)
+	if err != nil {
+		return nil, err
+	}
+	runtime.store = store
+
+	// Set up containers/image
+	runtime.imageContext = &types.SystemContext{
+		SignaturePolicyPath: runtime.config.SignaturePolicyPath,
+	}
+
+	runtime.seccompEnabled = seccomp.IsEnabled()
+	runtime.apparmorEnabled = apparmor.IsEnabled()
+
+	// Mark the runtime as valid - ready to be used, cannot be modified
+	// further
+	runtime.valid = true
+
+	return runtime, nil
+}
+
+// GetConfig returns a copy of the configuration used by the runtime
+func (r *Runtime) GetConfig() *RuntimeConfig {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	if !r.valid {
+		return nil
+	}
+
+	config := new(RuntimeConfig)
+
+	// Copy so the caller won't be able to modify the actual config
+	deepcopier.Copy(r.config).To(config)
+
+	return config
+}
+
+// Shutdown shuts down the runtime and associated containers and storage
+// If force is true, containers and mounted storage will be shut down before
+// cleaning up; if force is false, an error will be returned if there are
+// still containers running or mounted
+func (r *Runtime) Shutdown(force bool) error {
+	r.lock.Lock()
```

Comment:
Could we ever get into a deadlock here and not be able to shutdown?  I think this is OK as is, but thought I'd throw the question out there for consideration.  IDK if we might need a "super" shutdown that is triggered after waiting some configurable time or not.  I don't think that's needed, but don't know that for a fact.

Reply:
I can definitely see the value of having a super-force flag that overrides the locking requirement and shuts everything down to ensure we can at least gracefully shutdown if we hang somewhere. I think we can hold off doing so for now, though - we don't actually have anything that can hang yet